pgjdbc/src/main/java/org/postgresql/core/v3/BatchedQueryDecorator.java

```diff
@@ -0,0 +1,288 @@
+  }
+
+  /**
+   * Reset the batched query for next use.
+   */
+  public void reset() {
+    super.setStatementTypes(originalPreparedTypes);
+    resetBatchedCount();
+    length = 0;
+  }
+
+  @Override
+  public boolean isStatementReWritableInsert() {
+    return true;
+  }
+
+  /**
+   * The original meta data may need updating.
+   */
+  @Override
+  public void setStatementTypes(int[] types) {
+    super.setStatementTypes(types);
+    if (isOriginalStale(types)) {
+      updateOriginal(types);
+    }
+  }
+
+  /**
+   * Get the statement types for all parameters in the batch.
+   *
+   * @return int an array of {@link org.postgresql.core.Oid} parameter types
+   */
+  @Override
+  public int[] getStatementTypes() {
+    int types[] = super.getStatementTypes();
+    if (isOriginalStale(types)) {
+      /*
+       * Use opportunity to update originals if a ParameterDescribe has been
+       * called.
+       */
+      updateOriginal(types);
+    }
+    return resizeTypes();
+  }
+
+  /**
+   * Check fields and update if out of sync
+   *
+   * @return int an array of {@link org.postgresql.core.Oid} parameter types
+   */
+  private int[] resizeTypes() {
+    // provide types depending on batch size, which may vary
+    int expected = getBindPositions();
+    int[] types = super.getStatementTypes();
+    int before = 0;
+
+    if (types == null) {
+      types = fill(expected);
+    }
+    if (types.length < expected) {
+      before = types.length;
+      types = fill(expected);
+    }
+    if (before != types.length) {
+      setStatementTypes(types);
+    }
+    return types;
+  }
+
+  private int[] fill(int expected) {
+    int[] types = Arrays.copyOf(originalPreparedTypes, expected);
+    for (int row = 1; row < getBatchSize(); row += 1) {
+      System.arraycopy(originalPreparedTypes, 0, types, row
+          * originalPreparedTypes.length, originalPreparedTypes.length);
+    }
+    return types;
+  }
+
+  @Override
+  boolean isPreparedFor(int[] paramTypes) {
+    resizeTypes();
+    return isStatementParsed() && super.isPreparedFor(paramTypes);
+  }
+
+  @Override
+  byte[] getEncodedStatementName() {
+    if (batchedEncodedName == null) {
+      String n = super.getStatementName();
+      if (n != null) {
+        batchedEncodedName = Utils.encodeUTF8(n);
+      }
+    }
+    return batchedEncodedName;
+  }
+
+  @Override
+  void setStatementName(String statementName) {
+    if (statementName == null) {
+      batchedEncodedName = null;
+      super.setStatementName(null);
+    } else {
+      super.setStatementName(statementName);
+      batchedEncodedName = Utils.encodeUTF8(statementName);
+    }
+  }
+
+  /**
+   * Detect when the vanilla prepared type meta data is out of date.
+   *
+   * @param preparedTypes
+   *          meta data to compare with
+   * @return boolean value indicating if internal type information needs
+   *         updating
+   */
+  private boolean isOriginalStale(int[] preparedTypes) {
```

Comment:
Comments added to method explaining why. Added comment to next method indicating how.